tests/integration/test_hwp_pid_agent_integration.py

```diff
@@ -24,67 +25,30 @@ def test_testing(wait_for_crossbar):
 
 
 @pytest.mark.integtest
-def test_hwp_rotation_failed_connection_pid(wait_for_crossbar, run_agent_idle, client):
-    resp = client.init_connection.start()
-    print(resp)
-    # We can't really check anything here, the agent's going to exit during the
-    # init_conneciton task because it cannot connect to the PID controller.
-
-
-@pytest.mark.integtest
-def test_hwp_rotation_get_direction(wait_for_crossbar, pid_emu, run_agent, client):
-    responses = {'*R02': 'R02400000\r'}
-    pid_emu.define_responses(responses)
-
-    client.init_connection.wait()  # wait for connection to be made
-    resp = client.get_direction()
-    print(resp)
-    assert resp.status == ocs.OK
-    print(resp.session)
-    assert resp.session['op_code'] == OpCode.SUCCEEDED.value
-
-    # Test when in reverse
-    responses = {'*W02401388': 'W02\r',
-                 '*R02': 'R02401388\r'}
-    pid_emu.define_responses(responses)
-
-    client.set_direction(direction='1')
-    resp = client.get_direction()
-    print(resp)
-    assert resp.status == ocs.OK
-    print(resp.session)
-    assert resp.session['op_code'] == OpCode.SUCCEEDED.value
+def test_hwp_rotation_get_state(wait_for_crossbar, hwp_emu, run_agent, client):
+    resp = client.get_state()
+    state = resp.session['data']
+    print(state)
+    assert len(state.keys()) > 0
 
 
 @pytest.mark.integtest
-def test_hwp_rotation_set_direction(wait_for_crossbar, pid_emu, run_agent, client):
-    responses = {'*W02400000': 'W02\r',
-                 '*W02401388': 'W02\r'}
-    pid_emu.define_responses(responses)
-
-    client.init_connection.wait()  # wait for connection to be made
```

Comment:
This definition of the responses that the DeviceEmulator returns directly within the tests was an intended feature. It makes clear the intended communication between the agent and the emulator right up front.

I'd prefer this over burying the actual message and response within some other structure like the HWPEmulator, strictly for clarity when reading the test.

Since the tests have historically been written after the Agents, most commonly when I developed these I'd be running the test with no response configured locally, seeing what messages are being sent, and reverse engineer what a valid expected response would be, sometimes with access to the hardware such that I can directly test the message/response pattern.

It looks like the task count decreased, with the exception of adding `get_state`, so I would have thought most of these already defined responses would have worked, especially since the driver code didn't change. Was that not the case?

Reply:
I've gone back and forth on this in thinking again about it. The thing I like about directly defining the responses like this within the test is that it clearly defines the state of the device within the test.

That said, our tests are largely independent of the state of the device, as most tasks just store a return value (if any at all) and those return values aren't checked in the tests. But if they were being checked, depending on the state of the emulator complicates the tests. 

For example, if you checked the current frequency after changing the set point in the PID agent, the emulator spends some time getting there, rather than just directly getting the value we defined here in the test.

But, again, since our tests don't really check these values, perhaps my point is moot. Marking as resolved.